### PR TITLE
Set location metadata in uplink messages forwarded by the AS

### LINKED ISSUE #2669: Device location not included in metadata
#### Summary
Device latitude, longitude & altitude are not included in metadata received in the MQTT integration (and presumably other integrations)

#### Steps to Reproduce
1. Register device and set location in console
2. Create an MQTT integration
3. Send uplink
4. Note that location information is not included in message received in MQTT client

#### What do you see now?
Message as described in documentation - https://thethingsstack.io/v3.8.1/integrations/data-formats/

#### What do you want to see instead?
Similar to V2. Include device latitude, longitude and altitude in metadata

...

#### Environment
3.8.2 PAYG instance. Node-RED MQTT Client

#### How do you propose to implement this?
Include fields for device location

#### How do you propose to test this?
We'll see Lat/Lon fields in metadata

#### Can you do this yourself and submit a Pull Request?
Unfortunately not

### PR DESCRIPTION
**Pull request by neoaggelos**
#### Summary

Closes #2669 

#### Changes

- Add `locations` field in `ApplicationUplink` messages.
- Set field when handling uplink message in the AS.

#### Testing

- Tested locally, working on unit tests.

#### Checklist

- [X] Scope: The referenced issue is addressed, there are no unrelated changes.
- [X] Compatibility: The changes are backwards compatible with existing API, storage, configuration and CLI, according to the compatibility commitments in `README.md` for the chosen target branch.
- [ ] Documentation: Relevant documentation is added or updated.
- [x] Changelog: Significant features, behavior changes, deprecations and fixes are added to `CHANGELOG.md`.
- [X] Commits: Commit messages follow guidelines in `CONTRIBUTING.md`, there are no fixup commits left.

### CODE REVIEW

**Comment by neoaggelos** [comment]
Blocked by #3293

File: api/messages.proto
Comment (adriansmares) [inline api/messages.proto]: Why the `map` instead of `repeated` ? `Location` contains the source.

**Review by adriansmares — COMMENTED** [review]
Target your cache PR in order to distinguish the changes.

File: pkg/identityserver/end_device_registry.go
Comment (htdvisser) [inline pkg/identityserver/end_device_registry.go]: This should not be necessary. The IS already has middleware that assigns `ttnpb.AllClusterRights` when using cluster auth.

**Comment by johanstokking** [comment]
Lots of reviewers here; @neoaggelos who should do what?

File: cmd/ttn-lw-stack/commands/start.go
Comment (johanstokking) [inline cmd/ttn-lw-stack/commands/start.go]: We initialize the concrete implementations in a config package, not in the start command. See many other use cases.

File: cmd/ttn-lw-stack/commands/start.go
Comment (johanstokking) [inline cmd/ttn-lw-stack/commands/start.go]: Also don't bother with a cache implementation on this level; make `NewMemoryDeviceCache` responsible for that

**Comment by neoaggelos** [comment]
> Lots of reviewers here; @neoaggelos who should do what?

This is because of CODEOWNERS. This PR is currently blocked waiting for https://github.com/TheThingsNetwork/lorawan-stack/pull/3293

File: pkg/applicationserver/device_cache.go
Comment (johanstokking) [inline pkg/applicationserver/device_cache.go]: Since we're in `pkg/applicationserver`, don't embed this type as that exposes all methods of `gcache.Cache` on `memoryDeviceCache`, leading to future errors in the future.

File: pkg/applicationserver/applicationserver.go
Comment (johanstokking) [inline pkg/applicationserver/applicationserver.go]: `getCachedEndDevice()` implies that it uses a cache, but that is conditional here.

Better rename this to `fetchEndDevice()` and godoc that it may use the cache?

File: pkg/applicationserver/applicationserver.go
Comment (johanstokking) [inline pkg/applicationserver/applicationserver.go]: This is lots of magic; how does the caller know that only `locations` are fetched? That happens to be the first use case of getting the device from IS. But this should be passed as method argument.

**Review by johanstokking — CHANGES_REQUESTED** [review]
I'm not sure if I like the high coupling in AS this way. For inspiration, please see the tenant fetcher which has a cache.

You might want to change the cache interface to a generic "getter" with callback to fetch the entity if needed. The concrete implementation can then always use the callback to get the entity (no cache), use a memory cache, etc.

**Comment by johanstokking** [comment]
Wait, is https://github.com/TheThingsNetwork/lorawan-stack/pull/3293 the better version of this?

**Comment by neoaggelos** [comment]
> Wait, is #3293 the better version of this?

https://github.com/TheThingsNetwork/lorawan-stack/pull/3294#issuecomment-700526324. #3293 is the device fetcher extracted as a separate PR, so that this can focus on the proto changes instead.

**Comment by adriansmares** [comment]
> 
> 
> > Wait, is #3293 the better version of this?
> 
> [#3294 (comment)](https://github.com/TheThingsNetwork/lorawan-stack/pull/3294#issuecomment-700526324). #3293 is the device fetcher extracted as a separate PR, so that this can focus on the proto changes instead.

If you make such chains of PRs, please actually make them target one another instead of making it look like this PR introduces changes that it doesn't. While you indeed added the `blocking`/`blocks` labels and comments, the boundaries between the PRs are unclear and as such the reviewers do not actually know which change in the code belongs which. 

If you would have targeted your branch from #3293 here, instead of `3.10`, only the API protos changes would have been visible (exactly what you wanted).

GitHub would have also automatically changed the target of this branch back to `3.10` once #3293 was merged, so you don't end up with lingering references.

File: api/messages.proto
Comment (neoaggelos) [reply api/messages.proto]: This is the same format as in `EndDevice` https://github.com/TheThingsNetwork/lorawan-stack/blob/v3.10/api/end_device.proto#L437.

File: api/messages.proto
Comment (neoaggelos) [inline api/messages.proto:213]: This is to account for https://github.com/TheThingsNetwork/lorawan-stack/pull/3298/files#diff-5e41eaf89519c9fe313dc525c2d3712dR207

Comment (johanstokking) [reply api/messages.proto:213]: This link doesn't seem to go to a related target, can you elaborate?

Comment (htdvisser) [reply api/messages.proto:213]: I guess he meant the `google.protobuf.Duration consumed_airtime = 13` added in #3298.

**Review by htdvisser — APPROVED** [review]

**Review by johanstokking — APPROVED** [review]

File: pkg/applicationserver/applicationserver.go
Comment (johanstokking) [inline pkg/applicationserver/applicationserver.go]: Nit; we have a `logger` here already

**Review by johanstokking — APPROVED** [review]

**Review by adriansmares — APPROVED** [review]

**Pull request merged**
